// @wordpress/components/build/date-time/date/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePicker = DatePicker;
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _moment = _interopRequireDefault(require("moment"));

require("react-dates/initialize");

var _DayPickerSingleDateController = _interopRequireDefault(require("react-dates/lib/components/DayPickerSingleDateController"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _utils = require("./utils");

var _styles = require("./styles");

/**
 * External dependencies
 */
// Needed to initialise the default datepicker styles.
// See: https://github.com/airbnb/react-dates#initialize
// `react-dates` doesn't tree-shake correctly, so we import from the individual
// component here.

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const TIMEZONELESS_FORMAT = 'YYYY-MM-DDTHH:mm:ss';
const ARIAL_LABEL_TIME_FORMAT = 'dddd, LL';

const noop = () => {};

function DatePickerDay(_ref) {
  let {
    day,
    events = []
  } = _ref;
  const ref = (0, _element.useRef)(null);
  /*
   * a11y hack to make the `There is/are n events` string
   * available speaking for readers,
   * re-defining the aria-label attribute.
   * This attribute is handled by the react-dates component.
   */

  (0, _element.useEffect)(() => {
    var _ref$current;

    // Bail when no parent node.
    if (!((ref === null || ref === void 0 ? void 0 : (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.parentNode) instanceof Element)) {
      return;
    }

    const {
      parentNode
    } = ref.current;
    const dayAriaLabel = (0, _moment.default)(day).format(ARIAL_LABEL_TIME_FORMAT);

    if (!events.length) {
      // Set aria-label without event description.
      parentNode.setAttribute('aria-label', dayAriaLabel);
      return;
    }

    const dayWithEventsDescription = (0, _i18n.sprintf)( // translators: 1: Calendar day format, 2: Calendar event number.
    (0, _i18n._n)('%1$s. There is %2$d event.', '%1$s. There are %2$d events.', events.length), dayAriaLabel, events.length);
    parentNode.setAttribute('aria-label', dayWithEventsDescription);
  }, [day, events.length]);
  return (0, _element.createElement)(_styles.Day, {
    ref: ref,
    className: "components-datetime__date__day" // Unused, for backwards compatibility.
    ,
    hasEvents: !!(events !== null && events !== void 0 && events.length),
    alignment: "center"
  }, day.format('D'));
}
/**
 * DatePicker is a React component that renders a calendar for date selection.
 *
 * ```jsx
 * import { DatePicker } from '@wordpress/components';
 * import { useState } from '@wordpress/element';
 *
 * const MyDatePicker = () => {
 *   const [ date, setDate ] = useState( new Date() );
 *
 *   return (
 *     <DatePicker
 *       currentDate={ date }
 *       onChange={ ( newDate ) => setDate( newDate ) }
 *     />
 *   );
 * };
 * ```
 */


function DatePicker(_ref2) {
  let {
    currentDate,
    onChange,
    events,
    isInvalidDate,
    onMonthPreviewed,
    startOfWeek = 0
  } = _ref2;
  const nodeRef = (0, _element.useRef)(null);

  const onMonthPreviewedHandler = newMonthDate => {
    onMonthPreviewed === null || onMonthPreviewed === void 0 ? void 0 : onMonthPreviewed(newMonthDate.toISOString());
    keepFocusInside();
  };
  /*
   * Todo: We should remove this function ASAP.
   * It is kept because focus is lost when we click on the previous and next month buttons.
   * This focus loss closes the date picker popover.
   * Ideally we should add an upstream commit on react-dates to fix this issue.
   */


  const keepFocusInside = () => {
    if (!nodeRef.current) {
      return;
    }

    const {
      ownerDocument
    } = nodeRef.current;
    const {
      activeElement
    } = ownerDocument; // If focus was lost.

    if (!activeElement || !nodeRef.current.contains(ownerDocument.activeElement)) {
      // Retrieve the focus region div.
      const focusRegion = nodeRef.current.querySelector('.DayPicker_focusRegion');

      if (!(focusRegion instanceof HTMLElement)) {
        return;
      } // Keep the focus on focus region.


      focusRegion.focus();
    }
  };

  const onChangeMoment = newDate => {
    if (!newDate) {
      return;
    } // If currentDate is null, use now as momentTime to designate hours, minutes, seconds.


    const momentDate = currentDate ? (0, _moment.default)(currentDate) : (0, _moment.default)();
    const momentTime = {
      hours: momentDate.hours(),
      minutes: momentDate.minutes(),
      seconds: 0
    };
    onChange === null || onChange === void 0 ? void 0 : onChange(newDate.set(momentTime).format(TIMEZONELESS_FORMAT)); // Keep focus on the date picker.

    keepFocusInside();
  };

  const getEventsPerDay = day => {
    if (!(events !== null && events !== void 0 && events.length)) {
      return [];
    }

    return events.filter(eventDay => day.isSame(eventDay.date, 'day'));
  };

  const momentDate = (0, _utils.getMomentDate)(currentDate);
  return (0, _element.createElement)("div", {
    className: "components-datetime__date",
    ref: nodeRef
  }, (0, _element.createElement)(_DayPickerSingleDateController.default, {
    date: momentDate,
    initialVisibleMonth: null,
    daySize: 30,
    horizontalMonthPadding: 0,
    focused: true,
    hideKeyboardShortcutsPanel: true // This is a hack to force the calendar to update on month or year change
    // https://github.com/airbnb/react-dates/issues/240#issuecomment-361776665
    ,
    key: `datepicker-controller-${momentDate ? momentDate.format('MM-YYYY') : 'null'}`,
    noBorder: true,
    numberOfMonths: 1,
    onDateChange: onChangeMoment,
    transitionDuration: 0,
    weekDayFormat: "ddd",
    dayAriaLabelFormat: ARIAL_LABEL_TIME_FORMAT,
    isRTL: (0, _i18n.isRTL)(),
    isOutsideRange: date => {
      return !!isInvalidDate && isInvalidDate(date.toDate());
    },
    firstDayOfWeek: startOfWeek,
    onPrevMonthClick: onMonthPreviewedHandler,
    onNextMonthClick: onMonthPreviewedHandler,
    renderDayContents: day => (0, _element.createElement)(DatePickerDay, {
      day: day,
      events: getEventsPerDay(day)
    }),
    renderMonthElement: _ref3 => {
      let {
        month
      } = _ref3;
      return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)("strong", null, month.format('MMMM')), ' ', month.format('YYYY'));
    },
    renderNavPrevButton: _ref4 => {
      let {
        ariaLabel,
        ...props
      } = _ref4;
      return (0, _element.createElement)(_styles.NavPrevButton, (0, _extends2.default)({
        icon: _icons.arrowLeft,
        variant: "tertiary",
        "aria-label": ariaLabel
      }, props));
    },
    renderNavNextButton: _ref5 => {
      let {
        ariaLabel,
        ...props
      } = _ref5;
      return (0, _element.createElement)(_styles.NavNextButton, (0, _extends2.default)({
        icon: _icons.arrowRight,
        variant: "tertiary",
        "aria-label": ariaLabel
      }, props));
    },
    onFocusChange: noop
  }));
}

var _default = DatePicker;
exports.default = _default;
